src/daemon/ca/publishing.rs

```diff
@@ -1,232 +1,1223 @@
+}
+
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct OldKeyState {
+    current_set: CurrentKeyObjectSet,
+    old_set: BasicKeyObjectSet,
+}
+
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct CurrentKeyObjectSet {
+    #[serde(flatten)]
+    basic: BasicKeyObjectSet,
+    #[serde(with = "objects_to_roas_serde")]
+    roas: HashMap<ObjectName, PublishedRoa>,
+    #[serde(with = "objects_to_certs_serde")]
+    certs: HashMap<ObjectName, PublishedCert>,
+}
+
+impl CurrentKeyObjectSet {
     pub fn new(
+        signing_cert: RcvdCert,
         number: u64,
-        revocations_delta: RevocationsDelta,
-        manifest_info: ManifestInfo,
-        crl_info: CrlInfo,
-        objects_delta: ObjectsDelta,
+        revocations: Revocations,
+        manifest: PublishedManifest,
+        crl: PublishedCrl,
+        roas: HashMap<ObjectName, PublishedRoa>,
+        certs: HashMap<ObjectName, PublishedCert>,
     ) -> Self {
-        CurrentObjectSetDelta {
+        let basic = BasicKeyObjectSet {
+            signing_cert,
             number,
-            revocations_delta,
-            manifest_info,
-            crl_info,
-            objects_delta,
+            revocations,
+            manifest,
+            crl,
+            old_repo: None,
+        };
+        CurrentKeyObjectSet { basic, roas, certs }
+    }
+
+    #[allow(clippy::mutable_key_type)]
+    fn add_elements(&self, map: &mut HashMap<RepositoryContact, Vec<PublishElement>>, dflt_repo: &RepositoryContact) {
+        let repo = self.old_repo.as_ref().unwrap_or(dflt_repo);
+
+        let base_uri = self.signing_cert.ca_repository();
+        let mft_uri = base_uri.join(self.manifest.name().as_bytes());
+        let crl_uri = base_uri.join(self.crl.name().as_bytes());
+
+        let elements = map.entry(repo.clone()).or_insert(vec![]);
+        elements.push(PublishElement::new(Base64::from(&self.manifest.0), mft_uri));
+        elements.push(PublishElement::new(Base64::from(&self.crl.0), crl_uri));
+
+        for (name, roa) in &self.roas {
+            elements.push(PublishElement::new(
+                Base64::from(&roa.0),
+                base_uri.join(name.as_bytes()),
+            ));
+        }
+
+        for (name, cert) in &self.certs {
+            elements.push(PublishElement::new(
+                Base64::from(cert.as_ref()),
+                base_uri.join(name.as_bytes()),
+            ));
+        }
+    }
+
+    fn update_roas(
+        &mut self,
+        roa_updates: &RoaUpdates,
+        timing: &IssuanceTimingConfig,
+        signer: &KrillSigner,
+    ) -> KrillResult<()> {
+        for (name, roa) in roa_updates.added_roas()? {
+            if let Some(old) = self.roas.insert(name, roa) {
+                self.revocations.add(Revocation::from(&old));
+            }
+        }
+        for name in roa_updates.removed_roas() {
+            if let Some(old) = self.roas.remove(&name) {
```

Comment:
Likewise, isn't it odd here if removal attempts to remove a ROA which didn't exist in the set?